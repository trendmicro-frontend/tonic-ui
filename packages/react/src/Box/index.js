import styled from '@emotion/styled';
import isPropValid from '@emotion/is-prop-valid';
import { ensureArray } from 'ensure-type';
import { cx, system } from '../shared/styled-system';

const shouldForwardProp = (() => {
  const stylePropMap = ensureArray(system.propNames)
    .reduce((acc, val) => {
      acc[val] = true;
      return acc;
    }, {});
  const omittedStylePropMap = {
    ...stylePropMap,

    // The `as` prop is supported by Emotion
    'as': true,
  };

  return prop => isPropValid(prop) && !omittedStylePropMap[prop];
})();

const createPseudoClassTransformFunction = name => prop => {
  const result = {};
  if (Array.isArray(prop)) {
    result[`${name}(${prop[0]})`] = prop[1];
  } else if (typeof prop === 'object') {
    for (const key in prop) {
      if (Object.prototype.hasOwnProperty.call(prop, key)) {
        result[`${name}(${key})`] = prop[key];
      }
    }
  }
  return result;
};

/**
 * Pseudo-classes
 * https://developer.mozilla.org/en-US/docs/Web/CSS/
 */

const active = [
  '&:active',
  '&[data-active=true]'
].join(',');
const checked = [
  '&[aria-checked=true]',
  '&:checked',
].join(',');
const selected = [
  '&[aria-selected=true]',
  '&[data-selected=true]',
  '&:active[data-selected=true]',
  '&:hover[data-selected=true]',
].join(',');
const disabled = [
  '&[aria-disabled=true]',
  '&:disabled',
  '&:disabled:hover',
  '&:disabled:focus',
  '&:hover[aria-disabled=true]',
  '&:focus[aria-disabled=true]',
].join(',');
const empty = '&:empty';
const enabled = [
  '&:enabled',
  '&:enabled:focus',
  '&:enabled:hover',
].join(',');
const firstChild = '&:first-child';
const firstOfType = '&:first-of-type';
const fullscreen = '&:fullscreen';
const focus = '&:focus';
const focusActive = '&:focus:active';
const focusHover = '&:focus:hover';
const focusWithin = '&:focus-within';
const focusSelected = ['&[aria-selected=true]:focus', '&[data-selected=true]:focus'].join(',');
const hover = '&:hover';
const indeterminate = '&:indeterminate';
const valid = [
  '&[aria-invalid=false]',
  '&:valid',
].join(',');
const invalid = [
  '&[aria-invalid=true]',
  '&:invalid',
].join(',');
const lastChild = '&:last-child';
const lastOfType = '&:last-of-type';
const nthOfTypeFn = createPseudoClassTransformFunction('&:nth-of-type');
const readOnly = [
  '&[aria-readonly=true]',
  '&:read-only',
].join(',');
const visited = '&:visited';

/**
 * Pseudo-elements
 * https://developer.mozilla.org/en-US/docs/Web/CSS/
 */

const after = '&::after';
const backdrop = '&::backdrop';
const before = '&::before';
const cue = '&::cue';
const firstLetter = '&::first-letter';
const firstLine = '&::first-line';
const placeholder = '&::placeholder';
const selection = '&::selection';

const _Box = styled('div', {
  shouldForwardProp,
})(system);

const Box = styled(_Box)(
  ({
    // pseudo-classes
    _active,
    _checked,
    _selected,
    _disabled,
    _empty,
    _enabled,
    _firstChild,
    _firstOfType,
    _fullscreen,
    _focus,
    _focusActive,
    _focusHover,
    _focusWithin,
    _focusSelected,
    _hover,
    _indeterminate,
    _valid,
    _invalid,
    _lastChild,
    _lastOfType,
    _nthOfType,
    _readOnly,
    _visited,

    // pseudo-elements
    __after,
    __backdrop,
    __before,
    __cue,
    __firstLetter,
    __firstLine,
    __placeholder,
    __selection,
  }) => {
    let rest = null;

    if (_nthOfType) {
      rest = { ...rest, ...nthOfTypeFn(_nthOfType) };
    }

    return cx({
      /**
       * Pseudo-classes must be declared in a specific order, as shown below:
       *
       * ```
       * :link
       * :visited
       * :hover
       * :active
       * ```
       *
       * Each pseudo-class corresponds to an event which can only happen later in the timeline than the one before.
       *
       * That is to say:
       *
       * 1. A link is unvisited before it is visited.
       * 2. A link is visited before it is hovered over.
       * 3. A link is hovered over before it is in active use.
       */
      [focus]: _focus,
      [visited]: _visited,
      [hover]: _hover,
      [focusHover]: _focusHover,
      [active]: _active,
      [focusActive]: _focusActive,
      [focusSelected]: _focusSelected,
      [focusWithin]: _focusWithin,
      [checked]: _checked,
      [selected]: _selected,
      [disabled]: _disabled,
      [empty]: _empty,
      [enabled]: _enabled,
      [firstChild]: _firstChild,
      [firstOfType]: _firstOfType,
      [fullscreen]: _fullscreen,
      [indeterminate]: _indeterminate,
      [valid]: _valid,
      [invalid]: _invalid,
      [lastChild]: _lastChild,
      [lastOfType]: _lastOfType,
      [readOnly]: _readOnly,

      // pseudo-elements
      [after]: __after,
      [backdrop]: __backdrop,
      [before]: __before,
      [cue]: __cue,
      [firstLetter]: __firstLetter,
      [firstLine]: __firstLine,
      [placeholder]: __placeholder,
      [selection]: __selection,

      ...rest
    });
  },
);

Box.displayName = 'Box';

export default Box;
