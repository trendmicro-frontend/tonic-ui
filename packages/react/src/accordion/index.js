import Accordion from './Accordion';
import AccordionBody from './AccordionBody';
import AccordionCollapse from './AccordionCollapse';
import AccordionHeader from './AccordionHeader';
import AccordionItem from './AccordionItem';
import AccordionToggle from './AccordionToggle';
import AccordionToggleIcon from './AccordionToggleIcon';
import useAccordion from './useAccordion';
import useAccordionItem from './useAccordionItem';

Accordion.Body = AccordionBody;
Accordion.Collapse = AccordionCollapse;
Accordion.Header = AccordionHeader;
Accordion.Item = AccordionItem;
Accordion.Toggle = AccordionToggle;
Accordion.Toggle.Icon = AccordionToggleIcon;

export {
  Accordion,
  AccordionBody,
  AccordionCollapse,
  AccordionHeader,
  AccordionItem,
  AccordionToggle,
  AccordionToggleIcon,
  useAccordion,
  useAccordionItem,
};
