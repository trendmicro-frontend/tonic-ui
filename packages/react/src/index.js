// components
export * from './accordion';
export * from './alert';
export * from './badge';
export * from './box';
export * from './button';
export * from './checkbox';
export * from './color-mode';
export * from './color-style';
export * from './css-baseline';
export * from './divider';
export * from './drawer';
export * from './flat-button';
export * from './flex';
export * from './grid';
export * from './icon';
export * from './image';
export * from './input';
export * from './link';
export * from './menu';
export * from './modal';
export * from './pagination';
export * from './popover';
export * from './popper';
export * from './portal';
export * from './presence';
export * from './radio';
export * from './search-input';
export * from './scrollbar';
export * from './select';
export * from './skeleton';
export * from './space';
export * from './spinner';
export * from './stack';
export * from './switch';
export * from './table';
export * from './tabs';
export * from './tag';
export * from './text';
export * from './textarea';
export * from './theme';
export * from './toast';
export * from './tooltip';
export * from './transitions';
export * from './visually-hidden';

// hooks
export useDisclosure from './deprecated/useDisclosure'; // deprecated

// settings
export colorStyle from './shared/color-style';
export theme from './shared/theme';
