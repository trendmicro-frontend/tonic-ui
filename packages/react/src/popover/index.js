import Popover from './Popover';
import PopoverBody from './PopoverBody';
import PopoverContent from './PopoverContent';
import PopoverFooter from './PopoverFooter';
import PopoverHeader from './PopoverHeader';
import PopoverTrigger from './PopoverTrigger';

Popover.Header = PopoverHeader;
Popover.Body = PopoverBody;
Popover.Content = PopoverContent;
Popover.Footer = PopoverFooter;
Popover.Header = PopoverHeader;
Popover.Trigger = PopoverTrigger;

export {
  Popover,
  PopoverBody,
  PopoverContent,
  PopoverFooter,
  PopoverHeader,
  PopoverTrigger,
};
