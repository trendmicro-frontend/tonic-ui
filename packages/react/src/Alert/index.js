import React, { forwardRef } from 'react';
import Box from '../Box';
import ButtonBase from '../ButtonBase';
import Flex from '../Flex';
import Icon from '../Icon';
import Space from '../Space';
import {
  useAlertStyle,
  useAlertIconStyle,
  useAlertMessageStyle,
  useAlertCloseButtonStyle,
} from './styles';

const defaultVariant = 'solid';
const defaultSeverity = 'success';

const getIconBySeverity = (severity) => {
  const iconName = {
    success: 'success',
    info: 'info',
    warning: 'warning-triangle',
    error: 'error',
  }[severity];

  if (!iconName) {
    return null;
  }

  return (
    <Icon icon={`${iconName}`} />
  );
};

const AlertIcon = (props) => (
  <Flex {...props} />
);

const AlertMessage = (props) => (
  <Box {...props} />
);

const AlertCloseButton = (props) => (
  <ButtonBase {...props} />
);

const Alert = forwardRef((
  {
    isClosable: _isClosable = false,
    isCloseButtonVisible: LEGACY_isCloseButtonVisible = false, // eslint-disable-line camelcase
    onClose,
    variant = defaultVariant,
    severity = defaultSeverity,
    icon,
    children,
    ...rest
  },
  ref,
) => {
  const isClosable = _isClosable || LEGACY_isCloseButtonVisible; // eslint-disable-line camelcase
  const styleProps = useAlertStyle({ variant, severity });
  const iconStyleProps = useAlertIconStyle({ variant, severity });
  const messageStyleProps = useAlertMessageStyle();
  const closeButtonStyleProps = useAlertCloseButtonStyle({ variant });

  if (typeof icon === 'string') {
    icon = (<Icon icon={icon} />);
  }
  if (typeof icon === 'undefined') {
    icon = getIconBySeverity(severity);
  }

  return (
    <Flex
      ref={ref}
      align="flex-start"
      justify="space-between"
      {...styleProps}
      {...rest}
    >
      {!!icon && (
        <>
          <AlertIcon {...iconStyleProps}>
            {icon}
          </AlertIcon>
          <Space minWidth="2x" />
        </>
      )}
      <AlertMessage {...messageStyleProps}>
        {children}
      </AlertMessage>
      {!!isClosable && (
        <>
          <Space minWidth="4x" />
          <AlertCloseButton {...closeButtonStyleProps} onClick={onClose}>
            <Icon icon="close-s" />
          </AlertCloseButton>
        </>
      )}
    </Flex>
  );
});

Alert.displayName = 'Alert';

export default Alert;
