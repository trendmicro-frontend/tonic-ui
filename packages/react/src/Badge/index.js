import React, { forwardRef } from 'react';
import useBadgeStyle from './styles';
import Box from '../Box';

const Badge = forwardRef(
  (
    {
      variant = 'badge',
      variantColor = 'red',
      badgeContent,
      isHidden = false,
      children,
      offset,
      dotSize = 8,
      borderColor,
      borderWidth = 1,
      ...restProps
    },
    ref,
  ) => {
    const badgeBorderColor = borderColor ?? variantColor;
    const badgeStyleProps = useBadgeStyle({
      color: variantColor,
      hasChildren: !!children,
      showAsDot: variant === 'dot',
      offset,
      dotSize,
      borderColor: badgeBorderColor,
      borderWidth
    });

    return (
      <Box
        as="span"
        position="relative"
        display="inline-block"
      >
        {children}
        {!isHidden && (
          <Box
            ref={ref}
            as="span"
            fontWeight="normal"
            borderRadius="sm"
            whiteSpace="nowrap"
            {...badgeStyleProps}
            {...restProps}
          >
            {badgeContent}
          </Box>
        )}
      </Box>
    );
  }
);

Badge.displayName = 'Badge';

export default Badge;
