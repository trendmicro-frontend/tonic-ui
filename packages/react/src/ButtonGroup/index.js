import memoize from 'micro-memoize';
import React, { forwardRef } from 'react';
import Box from '../Box';
import { ButtonGroupProvider } from './context';

const getMemoizedState = memoize(state => ({ ...state }));

const ButtonGroup = forwardRef((
  {
    children,
    size = 'md',
    variant = 'default',
    orientation = 'horizontal',
    ...rest
  },
  ref
) => {
  const buttonGroupState = getMemoizedState({ size, variant, orientation });
  const orientationProps = {
    vertical: {
      flexDirection: 'column',
    },
    horizontal: {
      flexDirection: 'row',
    },
  }[orientation];
  return (
    <ButtonGroupProvider value={buttonGroupState}>
      <Box
        ref={ref}
        display="inline-flex"
        {...orientationProps}
        {...rest}
      >
        {children}
      </Box>
    </ButtonGroupProvider>
  );
});

ButtonGroup.displayName = 'ButtonGroup';

export default ButtonGroup;
