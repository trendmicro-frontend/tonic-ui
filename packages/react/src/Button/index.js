import React, { forwardRef } from 'react';
import ButtonBase from '../ButtonBase';
import { useButtonGroup } from '../ButtonGroup/context';
import { getButtonGroupCSS, useButtonStyle } from './styles';

const defaultSize = 'md';
const defaultVariant = 'default';

const Button = forwardRef(
  (
    {
      as: Comp = 'button',
      type = 'button',
      borderRadius = 'sm',
      size,
      variant,
      children,
      css,
      ...rest
    },
    ref,
  ) => {
    const buttonGroupContext = useButtonGroup();
    let isInGroup = false;
    let useVertical = false;
    if (buttonGroupContext) {
      isInGroup = true;
      const {
        size: buttonGroupSize,
        variant: buttonGroupVariant,
        orientation,
      } = { ...buttonGroupContext };
      useVertical = (orientation === 'vertical');
      // - Use the inherited value from the button group
      // - Fallback to the default value if the value is null or undefined
      size = (buttonGroupSize ?? size) ?? defaultSize;
      variant = (buttonGroupVariant ?? variant) ?? defaultVariant;
    } else {
      // Use the default value if the value is null or undefined
      size = size ?? defaultSize;
      variant = variant ?? defaultVariant;
    }
    const buttonStyleProps = useButtonStyle({
      size,
      variant,
      borderRadius,
    });
    css = [
      isInGroup && getButtonGroupCSS({ useVertical }),
      { ...css }
    ];

    return (
      <ButtonBase
        ref={ref}
        as={Comp}
        type={type}
        borderRadius={borderRadius}
        css={css}
        {...buttonStyleProps}
        {...rest}
      >
        { children }
      </ButtonBase>
    );
  },
);

Button.displayName = 'Button';

export default Button;
