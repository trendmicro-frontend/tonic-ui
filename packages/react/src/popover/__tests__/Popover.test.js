import { act, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { render } from '@tonic-ui/react/test-utils/render';
import { testA11y } from '@tonic-ui/react/test-utils/accessibility';
import { Button, Popover, PopoverTrigger, PopoverContent } from '@tonic-ui/react/src';
import React from 'react';

describe('Popover', () => {
  it('should render correctly', async () => {
    const user = userEvent.setup();
    const renderOptions = {
      useCSSVariables: true,
    };
    const { container } = render((
      <Popover>
        <PopoverTrigger data-testid="popover-trigger">
          <Button>Hover me</Button>
        </PopoverTrigger>
        <PopoverContent data-testid="popover-content">
          This is a popover
        </PopoverContent>
      </Popover>
    ), renderOptions);

    const popoverTrigger = screen.getByTestId('popover-trigger');
    expect(popoverTrigger).toBeInTheDocument();

    await act(() => user.click(popoverTrigger));

    await waitFor(() => {
      const popoverContent = screen.getByTestId('popover-content');
      expect(popoverContent).toHaveTextContent('This is a popover');
    });

    expect(await screen.findByRole('dialog')).toBeInTheDocument();

    expect(container).toMatchSnapshot();

    await testA11y(container);
  });
});
