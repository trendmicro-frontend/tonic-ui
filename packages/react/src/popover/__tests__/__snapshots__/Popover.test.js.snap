// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should render correctly 1`] = `
.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-gray-60);
  border: var(--tonic-borders-1);
  color: var(--tonic-colors-white-emphasis);
  cursor: pointer;
  outline: 0;
  padding: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  border-radius: var(--tonic-radii-sm);
  padding-left: var(--tonic-space-3x);
  padding-right: var(--tonic-space-3x);
  -webkit-transition: background-color 200ms cubic-bezier(0.4, 0, 0.2, 1),border-color 200ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 200ms cubic-bezier(0.4, 0, 0.2, 1),color 200ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: background-color 200ms cubic-bezier(0.4, 0, 0.2, 1),border-color 200ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 200ms cubic-bezier(0.4, 0, 0.2, 1),color 200ms cubic-bezier(0.4, 0, 0.2, 1);
  min-height: var(--tonic-sizes-8x);
  font-size: var(--tonic-fontSizes-sm);
  line-height: var(--tonic-lineHeights-sm);
  border-color: var(--tonic-colors-transparent);
}

.emotion-0[aria-selected=true],
.emotion-0[data-selected] {
  background-color: var(--tonic-colors-gray-70);
  color: var(--tonic-colors-white-emphasis);
  pointer-events: none;
}

.emotion-0:focus,
.emotion-0[data-focus] {
  z-index: 1;
}

.emotion-0:focus:not(:active),
.emotion-0[data-focus]:not(:active) {
  border-color: var(--tonic-colors-blue-60);
  box-shadow: inset 0 0 0 .0625rem #1e5ede,inset 0 0 0 .125rem rgba(0, 0, 0, 1.0);
}

.emotion-0:hover,
.emotion-0[data-hover] {
  background-color: var(--tonic-colors-gray-50);
  z-index: 2;
}

.emotion-0:active,
.emotion-0[data-active] {
  background-color: var(--tonic-colors-gray-70);
}

.emotion-0:disabled,
.emotion-0:disabled:focus,
.emotion-0:disabled:hover,
.emotion-0[aria-disabled=true],
.emotion-0[aria-disabled=true]:focus,
.emotion-0[aria-disabled=true]:hover,
.emotion-0[data-disabled] {
  background-color: var(--tonic-colors-gray-60);
  color: var(--tonic-colors-white-emphasis);
  cursor: not-allowed;
  opacity: 0.28;
}

.emotion-2 {
  position: absolute;
  z-index: var(--tonic-zIndices-popover);
}

.emotion-4 {
  background-color: var(--tonic-colors-gray-80);
  color: var(--tonic-colors-white-primary);
  box-shadow: 0 2px 8px 0 rgba(0, 0, 0, 0.48),0 1px 2px 0 rgba(0, 0, 0, 0.16);
  border-width: 1px;
  font-size: var(--tonic-fontSizes-sm);
  line-height: var(--tonic-lineHeights-sm);
  padding: var(--tonic-space-3x);
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-radius: var(--tonic-radii-sm);
  max-width: 288px;
  outline: 0;
  opacity: 1;
  -webkit-transform: none;
  -moz-transform: none;
  -ms-transform: none;
  transform: none;
  -webkit-transition: opacity cubic-bezier(0.4, 0, 0.2, 1),-webkit-transform cubic-bezier(0.4, 0, 0.2, 1);
  transition: opacity cubic-bezier(0.4, 0, 0.2, 1),transform cubic-bezier(0.4, 0, 0.2, 1);
  transform-origin: top center;
}

.emotion-6[data-popper-placement^="top"] {
  position: absolute;
  bottom: 0;
}

.emotion-6[data-popper-placement^="top"]::before {
  content: "";
  border-top: 8px solid;
  border-left: calc(12px/2) solid transparent;
  border-right: calc(12px/2) solid transparent;
  -webkit-filter: drop-shadow(0 1px 1px rgba(0, 0, 0, 0.16));
  filter: drop-shadow(0 1px 1px rgba(0, 0, 0, 0.16));
  position: absolute;
  bottom: calc(8px * -1);
  -webkit-transform: translateX(-50%);
  -moz-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.emotion-6[data-popper-placement^="bottom"] {
  position: absolute;
  top: 0;
}

.emotion-6[data-popper-placement^="bottom"]::before {
  content: "";
  border-bottom: 8px solid;
  border-left: calc(12px/2) solid transparent;
  border-right: calc(12px/2) solid transparent;
  -webkit-filter: drop-shadow(0 -1px 1px rgba(0, 0, 0, 0.16));
  filter: drop-shadow(0 -1px 1px rgba(0, 0, 0, 0.16));
  position: absolute;
  top: calc(8px * -1);
  -webkit-transform: translateX(-50%);
  -moz-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.emotion-6[data-popper-placement^="left"] {
  position: absolute;
  right: 0;
}

.emotion-6[data-popper-placement^="left"]::before {
  content: "";
  border-left: 8px solid;
  border-top: calc(12px/2) solid transparent;
  border-bottom: calc(12px/2) solid transparent;
  -webkit-filter: drop-shadow(1px 0px 1px rgba(0, 0, 0, 0.16));
  filter: drop-shadow(1px 0px 1px rgba(0, 0, 0, 0.16));
  position: absolute;
  right: calc(8px * -1);
  -webkit-transform: translateY(-50%);
  -moz-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.emotion-6[data-popper-placement^="right"] {
  position: absolute;
  left: 0;
}

.emotion-6[data-popper-placement^="right"]::before {
  content: "";
  border-right: 8px solid;
  border-top: calc(12px/2) solid transparent;
  border-bottom: calc(12px/2) solid transparent;
  -webkit-filter: drop-shadow(-1px 0px 1px rgba(0, 0, 0, 0.16));
  filter: drop-shadow(-1px 0px 1px rgba(0, 0, 0, 0.16));
  position: absolute;
  left: calc(8px * -1);
  -webkit-transform: translateY(-50%);
  -moz-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

<div>
  <button
    aria-controls="@tonic-ui/react:Popover-1"
    aria-describedby="@tonic-ui/react:Popover-1"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="emotion-0 emotion-1"
    data-testid="popover-trigger"
    id="@tonic-ui/react:PopoverTrigger-1"
    type="button"
  >
    Hover me
  </button>
  <div
    aria-labelledby="@tonic-ui/react:PopoverTrigger-1"
    class="emotion-2 emotion-1"
    data-popper-escaped=""
    data-popper-placement="bottom"
    data-popper-reference-hidden=""
    id="@tonic-ui/react:Popover-1"
    role="dialog"
    style="position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 12px);"
  >
    <div
      class="emotion-4 emotion-1"
      data-testid="popover-content"
      tabindex="-1"
    >
      <div
        class="emotion-6 emotion-1"
        data-popper-arrow="true"
        data-popper-placement="bottom"
        role="presentation"
        style="position: absolute; left: 0px; transform: translate(12px, 0px);"
      />
      This is a popover
    </div>
  </div>
</div>
`;
